gonat_linux.go

```diff
@@ -0,0 +1,337 @@
+)
+
+type server struct {
+	acceptedPackets int64
+	rejectedPackets int64
+	numTCPConns     int64
+	numUDPConns     int64
+
+	tcpSocket          io.ReadWriteCloser
+	udpSocket          io.ReadWriteCloser
+	downstream         io.ReadWriter
+	opts               *Opts
+	bufferPool         BufferPool
+	ctrack             *ct.Conn
+	ctTimeout          uint32
+	randomPortSequence []uint16
+	portIndexes        map[uint8]map[Addr]int
+	connsByDownFT      map[FiveTuple]*conn
+	connsByUpFT        map[FiveTuple]*conn
+	fromDownstream     chan *IPPacket
+	toDownstream       chan *IPPacket
+	fromUpstream       chan *IPPacket
+	closedConns        chan *conn
+	close              chan interface{}
+}
+
+// NewServer constructs a new Server that reads packets from downstream
+// and writes response packets back to downstream.
+func NewServer(downstream io.ReadWriter, opts *Opts) (Server, error) {
+	err := opts.ApplyDefaults()
+	if err != nil {
+		return nil, errors.New("Error applying default options: %v", err)
+	}
+
+	log.Debugf("Outbound packets will use %v", opts.IFAddr)
+
+	ctrack, err := ct.Dial(nil)
+	if err != nil {
+		return nil, errors.New("Unable to obtain connection for managing conntrack: %v", err)
+	}
+	_ctTimeout := opts.IdleTimeout * 2
+	if _ctTimeout < MinConntrackTimeout {
+		_ctTimeout = MinConntrackTimeout
+	}
+	ctTimeout := uint32(_ctTimeout.Seconds())
+
+	// We create a random order for assigning new ports to minimize the chance of colliding
+	// with other running gonat instances.
+	randomPortSequence := make([]uint16, numEphemeralPorts)
+	for i := uint16(0); i < uint16(numEphemeralPorts); i++ {
+		randomPortSequence[i] = minEphemeralPort + i
+	}
+	rnd := rand.New(rand.NewSource(time.Now().UnixNano()))
+	rnd.Shuffle(numEphemeralPorts, func(i int, j int) {
+		randomPortSequence[i], randomPortSequence[j] = randomPortSequence[j], randomPortSequence[i]
+	})
+
+	s := &server{
+		downstream:         downstream,
+		opts:               opts,
+		bufferPool:         opts.BufferPool,
+		ctrack:             ctrack,
+		ctTimeout:          ctTimeout,
+		randomPortSequence: randomPortSequence,
+		portIndexes:        make(map[uint8]map[Addr]int),
+		connsByDownFT:      make(map[FiveTuple]*conn),
+		connsByUpFT:        make(map[FiveTuple]*conn),
+		fromDownstream:     make(chan *IPPacket, opts.BufferDepth),
+		toDownstream:       make(chan *IPPacket, opts.BufferDepth),
+		fromUpstream:       make(chan *IPPacket, opts.BufferDepth),
+		closedConns:        make(chan *conn, opts.BufferDepth),
+		close:              make(chan interface{}),
+	}
+	return s, nil
+}
+
+func (s *server) Serve() error {
+	var err error
+	s.tcpSocket, err = createSocket(FiveTuple{IPProto: syscall.IPPROTO_TCP, Src: Addr{s.opts.IFAddr, 0}})
+	if err != nil {
+		return err
+	}
+	ops.Go(func() { s.readFromUpstream(s.tcpSocket) })
+
+	s.udpSocket, err = createSocket(FiveTuple{IPProto: syscall.IPPROTO_UDP, Src: Addr{s.opts.IFAddr, 0}})
+	if err != nil {
+		return err
```

Comment:
Again, returning here leaves stuff open (now the conntrack.Conn object and the TCP socket).

Notably, even if the caller did something like:

```
if err := s.Serve(); err != nil {
    s.Close()
}
```

the resources would still be open as Close is a no-op if dispatch has not been called.

Reply:
Good catch, will fix.